dragonfly/block/farmland.go

```diff
@@ -0,0 +1,105 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/dragonfly/internal/item_internal"
+	"github.com/df-mc/dragonfly/dragonfly/item"
+	"github.com/df-mc/dragonfly/dragonfly/world"
+	"math/rand"
+)
+
+type Farmland struct {
+	// Hydration is how much moisture a block has. This is calculated by checking if there is a water source block
+	// 4 blocks away in any direction from the farmland that is either 1 block above or on the same level as the farmland block.
+	// If there isn't, we then count down the hydration level by one until it eventually dries up and turns back into dirt.
+	Hydration uint8
+}
+
+// NeighbourUpdateTick checks if a block was placed above the farmland that isn't a crop, turn the farmland into dirt.
+func (f Farmland) NeighbourUpdateTick(pos, block world.BlockPos, w *world.World) {
+	if _, isAir := w.Block(pos.Side(world.FaceUp)).(Air); !isAir {
+		if _, isCrop := w.Block(pos.Side(world.FaceUp)).(Crop); !isCrop {
+			w.SetBlock(pos, Dirt{})
+		}
+	}
+}
+
+// RandomTick wil make an attempt to grow crops, along with dehydrating the land and updating the hydration level.
+func (f Farmland) RandomTick(pos world.BlockPos, w *world.World, r *rand.Rand) {
+	// Calculate the Hydration of the farmland block
+	f.CalculateHydration(pos, w)
+	// Check if there is a crop on the farmland block.
+	if crop, isCrop := w.Block(pos.Add(world.BlockPos{0, 1})).(Crop); isCrop {
+		// Check if the crop requires Hydration to grow and if it meets those requirements.
+		if crop.RequiresHydration() && f.Hydration == 0 {
+			return
+		}
+
+		// Check if the crop can grow due to lighting
+		if w.Light(world.BlockPos{0, 1}) >= crop.LightLevelRequired() {
```

Comment:
Are you not meant to add {0, 1} to pos?

Reply:
Crops grow based on the lighting of the crop block itself. I added 1 to the position to get the location of the crop block for lighting

Reply:
You check if the light level at (0, 1, 0) is the right one, not the b,lock above the crop

Reply:
Hmm, yeah I see it now. That really slipped right through my fingers